### ISSUE TITLE
Instructions for `pkgs.k8s.io` should be updated to check GPG fingerprint

### ISSUE THREAD

**Issue by AkihiroSuda** [kubernetes/website#42718]
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/
https://kubernetes.io/blog/2023/08/15/pkgs-k8s-io-introduction/

The current documentations instruct to trust `https://pkgs.k8s.io/core:/stable:/v1.28/deb/Release.key` without checking its fingerprint:

https://github.com/kubernetes/website/blob/833291687fd2a241efd3cb295dc4973af3696942/content/en/docs/setup/production-environment/tools/kubeadm/install-kubeadm.md?plain=1#L181-L192

This doesn't seem to be more secure than just relying on HTTPS.
The documentation should be updated to check its fingerprint.
Or, the GPG can be just removed if the HTTPS infrastructure for `https://pkgs.k8s.io` is considered to be trustable.

**Comment by sftim** [comment]
/sig security

/priority awaiting-more-evidence

Is there any useful way to verify that signing key?

**Comment by sftim** [comment]
> the HTTPS infrastructure for https://pkgs.k8s.io

That's on AWS now, but in the future we might serve packages from other places (eg using a redirect).

Options here include:
- publish the signing key to a URL that we consider even more trusted
- sign the package signing key using GPG and explain how to validate it
- sign the package signing key using some other method, eg S/MIME, and explain how to validate back to a trust authority

**Comment by afbjorklund** [comment]
The hosting of the key seems to be here: https://build.opensuse.org/projects/isv:kubernetes/signing_keys
> Fingerprint `de15 b144 86cd 377b 9e87 6e1a 2346 54da 9a29 6436`

But there were some issues with the _previous_ key, https://packages.cloud.google.com/apt/doc/apt-key.gpg
that caused it to be [moved](https://github.com/kubernetes/website/commit/70266d5f79db1fc8f3db7bc04571bf0725789b05) from a different server to the regular dl.k8s.io infrastructure (like the `stable.txt`)

**Comment by xmudrii** [comment]
I'll be taking a look into this.
/assign

**Comment by afbjorklund** [comment]
Something simple like hardcoding the fingerprint in the install script and verifying against the downloaded key ?

`gpg --show-keys kubernetes-apt-keyring.gpg | grep --quiet DE15B14486CD377B9E876E1A234654DA9A296436`

**Comment by AkihiroSuda** [comment]
> Something simple like hardcoding the fingerprint in the install script and verifying against the downloaded key ?
> 
> `gpg --show-keys kubernetes-apt-keyring.gpg | grep --quiet DE15B14486CD377B9E876E1A234654DA9A296436`
> 
> "[expires: 2024-11-02]"

This grep seems vulnerable: An attacker may put "DE15B14486CD377B9E876E1A234654DA9A296436" to the signer name, the email address, or maybe other metadata.

**Comment by afbjorklund** [comment]
The default format only has significant whitespace:

```
pub   rsa2048 2022-08-25 [SC] [expires: 2024-11-02]
      DE15B14486CD377B9E876E1A234654DA9A296436
uid                      isv:kubernetes OBS Project <isv:kubernetes@build.opensuse.org>

```

So either it something hacky like `^      `, or use the machine readable format.

```sh
gpg --show-keys --with-colons kubernetes-apt-keyring.gpg | grep --quiet "fpr:::::::::DE15B14486CD377B9E876E1A234654DA9A296436:"
```

----

It actually did say so in the manual:

> **Never use the output of this command in scripts or other programs.** The output is intended only for humans and its format is likely to change.  The --with-colons option emits the output in a stable, machine-parseable format, which is intended for use by scripts and other programs.

But since we're doing a simple install script anyway, might as well put  in the complicated instructions?

Note that the `containerd.io` package does not check, either: https://docs.docker.com/engine/install/ubuntu/

`curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg`

**Comment by cShirley14** [comment]
(Found this thread due to frustration with documentation around pgp keys.)

The following command seems more natural and secure: `gpg --fingerprint DE15B14486CD377B9E876E1A234654DA9A296436`

**Comment by sftim** [comment]
/lifecycle frozen

Let's make sure to triage this issue.

**Comment by afbjorklund** [comment]
The suggested command doesn't seem to help much:

```console
$ gpg --fingerprint DE15B14486CD377B9E876E1A234654DA9A296436
gpg: error reading key: No public key
```

And it is not obvious, where it should go in the docs...

----

i.e. how to check the downloaded key, before using?

OBS: [isv_kubernetes_key.gpg](https://build.opensuse.org/projects/isv:kubernetes/signing_keys/download?kind=gpg)
deb: [Release.key](https://pkgs.k8s.io/core:/stable:/v1.28/deb/Release.key)
rpm: [repomd.xml.key](https://pkgs.k8s.io/core:/stable:/v1.28/rpm/repodata/repomd.xml.key)

Same thing also goes for the add-ons, such as cri-o:

https://kubernetes.io/blog/2023/10/10/cri-o-community-package-infrastructure/
* https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/rpm/repodata/repomd.xml.key
* https://pkgs.k8s.io/addons:/cri-o:/prerelease:/main/deb/Release.key

**Comment by cShirley14** [comment]
FWIW, if you take a moment to re-read my response you’ll see I wasn’t attempting to solve your initial problem as much as I was responding to the attempt to “hack” a solution with grep and it’s inherent flaws. It makes more sense to me to use gpg, which already has functionality that will take a known Fingerprint and list it if it there. 

And not sure what you are talking about in terms of fingerprints and there being “no obvious” place to add in docs, I think you’re taking what I’m saying out of context… e.g. companies who list public fingerprints along with public keys right next to their install verification instructions. The use case seems pretty obvious to me ([vault](https://www.hashicorp.com/trust/security), [aws](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html)) 🤷

**Comment by afbjorklund** [comment]
> It makes more sense to me to use gpg, which already has functionality that will take a known Fingerprint and list it if it there.

Right, it was just missing some details like that one needs to import the key to the keyring first? Currently it is just stored in a file...

Listing the fingerprint sounds like a good idea. And I think the suggestion of an install script was already turned down.

So we're not doing any more hacks, but waiting for the real solution.

**Comment by sftim** [comment]
This is one for SIG Release and SIG Security. The best first step would be to move things forward so we can triage it, reaching consensus on whether anything needs fixing.

**Comment by xmudrii** [comment]
I'm not working on this at the moment.
/unassign
